File: Embarcadero/Berlin/CPP/NotesResource/NotesClientModuleU.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#pragma hdrstop
#include <memory>
#include "NotesClientModuleU.h"
#include <FMX.Dialogs.hpp>
#include <REST.Types.hpp>
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma classgroup "FMX.Controls.TControl"
#pragma resource "*.dfm"
TNotesClientModule *NotesClientModule;
//---------------------------------------------------------------------------
__fastcall TNotesClientModule::TNotesClientModule(TComponent* Owner)
	: TDataModule(Owner)
{
}
//---------------------------------------------------------------------------
TJSONArray * __fastcall TNotesClientModule::GetJSONArray(const TCustomRESTResponse *AResponse)
{
	if(const_cast<TCustomRESTResponse*>(AResponse)->JSONValue == NULL) {
		throw new System::Sysutils::Exception("JSON expected");
	}
	return static_cast<TJSONArray*>(const_cast<TCustomRESTResponse*>(AResponse)->JSONValue);
}
//---------------------------------------------------------------------------
TJSONObject * __fastcall TNotesClientModule::GetJSONObject(const TCustomRESTResponse *AResponse)
{
	if(const_cast<TCustomRESTResponse*>(AResponse)->JSONValue == NULL) {
		throw new System::Sysutils::Exception("JSON expected");
	}
	return static_cast<TJSONObject*>(const_cast<TCustomRESTResponse*>(AResponse)->JSONValue);
}
//---------------------------------------------------------------------------
bool __fastcall TNotesClientModule::GetLoggedIn(void)
{
	return BackendAuth1->LoggedIn;
}
//---------------------------------------------------------------------------
String __fastcall TNotesClientModule::GetLoggedInUserName(void)
{
	return BackendAuth1->LoggedInUserName;
}
 //---------------------------------------------------------------------------
void __fastcall TNotesClientModule::Login(const String & AUserName, const String & APassword)
{
	BackendAuth1->UserName = AUserName;
	BackendAuth1->Password = APassword;
	BackendAuth1->Login();
}
//---------------------------------------------------------------------------
void __fastcall TNotesClientModule::Logout(void)
{
	BackendAuth1->Logout();
}
//---------------------------------------------------------------------------
void __fastcall TNotesClientModule::Signup(const String & AUserName, const String & APassword)
{
	BackendAuth1->UserName = AUserName;
	BackendAuth1->Password = APassword;
	BackendAuth1->Signup();
}
//---------------------------------------------------------------------------
void __fastcall TNotesClientModule::AddNote(const TNote * ANote, String &AID)
{
	TBackendEndpoint * lEndpoint = BackendEndpointAddNote;
	std::auto_ptr<TJSONObject> ljsonObj(TNoteJSON::NoteToJSON(ANote));
	lEndpoint->ClearBody();
	lEndpoint->AddBody(ljsonObj.get());
	lEndpoint->Execute();
	AID = GetJSONObject(lEndpoint->Response)->Get("id")->JsonValue->Value();
}
//---------------------------------------------------------------------------
bool __fastcall TNotesClientModule::FindNote(const String & ATitle, TNote * ANote)
{
	TBackendEndpoint *lEndpoint = BackendEndpointGetNotes;
	TRESTRequestParameter * lParam = lEndpoint->Params->AddItem();
	std::vector<TNote*> lNotes;
	bool _return = false;
	try
	{
		// Add a parameter to get note request
		lParam->Kind = TRESTRequestParameterKind::pkGETorPOST;
		lParam->name = "title";
		lParam->Value = ATitle;
		BackendEndpointGetNote->Execute();
		lNotes = TNoteJSON::JSONToNotes(GetJSONArray(lEndpoint->Response));
		if(lNotes.size() == 0) {
			_return = false;
		}
		std::vector<TNote*>::iterator it = lNotes.begin();
		ANote = *it;
		_return = true;
    }
     __finally
	{
		lEndpoint->Params->Delete(lParam);
	}
	return _return;
}
//---------------------------------------------------------------------------
bool __fastcall TNotesClientModule::DeleteNote(const String &AID)
{
	TBackendEndpoint *lEndpoint = BackendEndpointDeleteNote;
	lEndpoint->Params->Items[0]->Value = AID;
	lEndpoint->AllowHTTPErrors = TBackendEndpoint::TAllowHTTPErrors::ClientErrorNotFound_404;
	lEndpoint->Execute();
	return lEndpoint->Response->Status.Success();
}
//---------------------------------------------------------------------------
bool __fastcall TNotesClientModule::GetNote(const String &AID, TNote *ANote)
{
	bool _return = false;
	TBackendEndpoint *lEndpoint = BackendEndpointGetNote;
	lEndpoint->Params->Items[0]->Value = AID;
	lEndpoint->AllowHTTPErrors = TBackendEndpoint::TAllowHTTPErrors::ClientErrorNotFound_404;
	lEndpoint->Execute();
	_return = lEndpoint->Response->Status.Success();
	if(_return) {
		ANote = TNoteJSON::JSONToNote(GetJSONObject(lEndpoint->Response));
	}
	return _return;
}
//---------------------------------------------------------------------------
void __fastcall TNotesClientModule::UpdateNote(const TNote * ANote)
{
	TBackendEndpoint * lEndpoint = BackendEndpointUpdateNote;
	lEndpoint->Params->Items[0]->Value = ANote->ID;
	std::auto_ptr<TJSONObject> ljsonObj(TNoteJSON::NoteToJSON(ANote));
	lEndpoint->ClearBody();
	lEndpoint->AddBody(ljsonObj.get());
	lEndpoint->Execute();
}
//---------------------------------------------------------------------------
std::vector<TNote*> __fastcall TNotesClientModule::GetNotes(void)
{
	TBackendEndpoint * lEndpoint = BackendEndpointGetNotes;
	lEndpoint->Execute();
	return TNoteJSON::JSONToNotes(GetJSONArray(lEndpoint->Response));
}
//---------------------------------------------------------------------------

File: Embarcadero/Rio/Object Pascal/NotesResource/NotesClientModuleU.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit NotesClientModuleU;

interface

uses
  System.SysUtils, System.Classes, IPPeerClient, REST.Backend.ServiceTypes,
  REST.Backend.MetaTypes, REST.Backend.EMSServices, System.JSON,
  REST.Backend.EMSProvider, REST.Client, REST.Backend.EndPoint,
  Data.Bind.Components, Data.Bind.ObjectScope, REST.Backend.BindSource,
  REST.Backend.ServiceComponents, NoteTypesU, System.Actions, FMX.ActnList;

type
  TNotesClientModule = class(TDataModule)
    BackendAuth1: TBackendAuth;
    BackendEndpointDeleteNote: TBackendEndpoint;
    BackendEndpointUpdateNote: TBackendEndpoint;
    BackendEndpointGetNotes: TBackendEndpoint;
    EMSProvider1: TEMSProvider;
    BackendEndpointGetNote: TBackendEndpoint;
    BackendEndpointAddNote: TBackendEndpoint;
  private
    function GetJSON<T>(const AResponse: TCustomRESTResponse): T;
    function GetJSONArray(const AResponse: TCustomRESTResponse): TJSONArray;
    function GetJSONObject(const AResponse: TCustomRESTResponse): TJSONObject;
    function GetLoggedIn: Boolean;
    function GetLoggedInUserName: string;
    { Private declarations }
  public
    { Public declarations }
    procedure Login(const AUserName, APassword: string);
    procedure Logout;
    procedure Signup(const AUserName, APassword: string);
    procedure AddNote(const ANote: TNote; out AID: string);
    function FindNote(const ATitle: string; out ANote: TNote): Boolean;
    function DeleteNote(const AID: string): Boolean;
    function GetNote(const AID: string; out ANote: TNote): Boolean;
    procedure UpdateNote(const ANote: TNote);
    function GetNotes: TArray<TNote>;
    property LoggedIn: Boolean read GetLoggedIn;
    property LoggedInUserName: string read GetLoggedInUserName;
  end;

var
  NotesClientModule: TNotesClientModule;

implementation

{%CLASSGROUP 'FMX.Controls.TControl'}

uses FMX.Dialogs, REST.Types;

{$R *.dfm}

function TNotesClientModule.GetJSON<T>(const AResponse: TCustomRESTResponse): T;
begin
  if AResponse.JSONValue = nil then
    raise Exception.Create('JSON Expected');
  Result := AResponse.JSONValue.GetValue<T>('');
end;

function TNotesClientModule.GetJSONObject(const AResponse: TCustomRESTResponse): TJSONObject;
begin
  Result := GetJSON<TJSONObject>(AResponse);
end;

function TNotesClientModule.GetLoggedIn: Boolean;
begin
  Result := BackendAuth1.LoggedIn;
end;

function TNotesClientModule.GetLoggedInUserName: string;
begin
  Result := BackendAuth1.LoggedInUserName;
end;

function TNotesClientModule.GetJSONArray(const AResponse: TCustomRESTResponse): TJSONArray;
begin
  Result := GetJSON<TJSONArray>(AResponse);
end;

procedure TNotesClientModule.AddNote(const ANote: TNote; out AID: string);
var
  LJSONObject: TJSONObject;
  LEndpoint: TBackendEndpoint;
begin
  LEndpoint := BackendEndpointAddNote;
  LJSONObject := TNoteJSON.NoteToJSON(ANote);
  try
    LEndpoint.ClearBody;
    LEndpoint.AddBody(LJSONObject);
    LEndpoint.Execute;
    AID := GetJSONObject(LEndpoint.Response).GetValue<string>('id');
  finally
    LJSONObject.Free;
  end;
end;

function TNotesClientModule.DeleteNote(const AID: string): Boolean;
var
  LEndpoint: TBackendEndpoint;
begin
  LEndpoint := BackendEndpointDeleteNote;
  LEndpoint.Params.Items[0].Value := AID;
  // Don't raise exception on 404
  LEndpoint.AllowHTTPErrors := TBackendEndpoint.TAllowHTTPErrors.ClientErrorNotFound_404;
  LEndpoint.Execute;
  Result := LEndpoint.Response.Status.Success;
end;

function TNotesClientModule.GetNote(const AID: string; out ANote: TNote): Boolean;
var
  LEndpoint: TBackendEndpoint;
begin
  LEndpoint := BackendEndpointGetNote;
  LEndpoint.Params.Items[0].Value := AID;
  // Don't raise exception on 404
  LEndpoint.AllowHTTPErrors := TBackendEndpoint.TAllowHTTPErrors.ClientErrorNotFound_404;
  LEndpoint.Execute;
  Result := LEndpoint.Response.Status.Success;
  if Result then
  begin
    ANote := TNoteJSON.JSONToNote(
      GetJSONObject(LEndpoint.Response));
  end;
end;

function TNotesClientModule.FindNote(const ATitle: string; out ANote: TNote): Boolean;
var
  LNotes: TArray<TNote>;
  LParam: TRESTRequestParameter;
  LEndpoint: TBackendEndpoint;
begin
  LEndpoint := BackendEndpointGetNotes;
  LParam := LEndpoint.Params.AddItem;
  try
    // Add a parameter to get note request
    LParam.Kind := TRESTRequestParameterKind.pkGETorPOST;
    LParam.Name := 'title';
    LParam.Value := ATitle;
    BackendEndpointGetNote.Execute;
    LNotes := TNoteJSON.JSONToNotes(
      GetJSONArray(LEndpoint.Response));
    if Length(LNotes) = 0 then
      Result := False
    else
    begin
      Assert(Length(LNotes) = 1);
      ANote := LNotes[0];
      Result := True;
    end;
  finally
    // Delete the parameter
    LEndpoint.Params.Delete(LParam);
  end;
end;

function TNotesClientModule.GetNotes: TArray<TNote>;
var
  LEndpoint: TBackendEndpoint;
begin
  LEndpoint := BackendEndpointGetNotes;
  LEndpoint.Execute;
  Result := TNoteJSON.JSONToNotes(
    GetJSONArray(LEndpoint.Response));
end;

procedure TNotesClientModule.Login(const AUserName, APassword: string);
begin
  BackendAuth1.UserName := AUserName;
  BackendAuth1.Password := APassword;
  BackendAuth1.Login;
end;

procedure TNotesClientModule.Logout;
begin
  BackendAuth1.Logout;
end;

procedure TNotesClientModule.Signup(const AUserName, APassword: string);
begin
  BackendAuth1.UserName := AUserName;
  BackendAuth1.Password := APassword;
  BackendAuth1.Signup;
end;

procedure TNotesClientModule.UpdateNote(const ANote: TNote);
var
  LJSONObject: TJSONObject;
  LEndpoint: TBackendEndpoint;
begin
  LEndpoint := BackendEndpointUpdateNote;
  LEndpoint.Params.Items[0].Value := ANote.ID;
  LJSONObject := TNoteJSON.NoteToJSON(ANote);
  try
    LEndpoint.ClearBody;
    LEndpoint.AddBody(LJSONObject);
    LEndpoint.Execute;
  finally
    LJSONObject.Free;
  end;
end;

end.
